### DISCUSSION TITLE
Pre-Building Generated Types and Marten.Events.EventGraph

### DISCUSSION THREAD

**Discussion by mikkelbu** [JasperFx/marten#2489] [Q&A]
Hi all

We have used Marten for a handful of years together with CQRSLite - https://github.com/gautema/CQRSlite - in a very simplistic setup of event sourcing (and I guess we have only used a very tiny fraction of Marten to do this - essentially we are just fetching the events using `IQueryEventStore.QueryAllRawEvents`, adding events using `IDocumentSession.Events.Append`, and persisting the events using `IDocumentSession.SaveChangesAsync`). And the event sourcing is just a little part of the services. 

We are in the process of migrating all our services from .NET core 2.1 to .NET 6, and in the process we are also bumping all our dependencies and thus we are bumping Marten from 3.13.4 to 5.10.1. We are in the process of testing the migrations and have discovered that our process are often getting restarted due to a high memory usage (normally before the migration the processes run at ~500 megabytes and the limit is set to 1 gigabyte). Now, after the migration we see that they use more and more memory until they are restarted. We have tried to set the limit to 4 gigabyte, but this is still not enough and it seem like the memory is never claimed by the GC, so the memory usage is just increasing steadily until the process is restarted.

As far as we can tell this is due to the dynamic creation of code using Roslyn - https://jeremydmiller.com/2015/11/11/using-roslyn-for-runtime-code-generation-in-marten/ - so we have tried to pre-build the relevant types and then include these in the project - using https://martendb.io/configuration/prebuilding.html.

When we set `GeneratedCodeMode` to create the code dynamically at runtime, then the attached file `EventStorage.cs` is created in `Internal\Generated\EventStore` (containing classes `GeneratedEventDocumentStorage`, `AppendEventOperation`, `GeneratedInsertStream`, `GeneratedStreamStateQueryHandler`,  and `GeneratedStreamVersionOperation`), but if we then add this file to the project and try to run the processes using `TypeLoadMode.Static`, then the code fails with the following stacktrace

```
LamarCompiler.ExpectedTypeMissingException: Could not load expected pre-built types for code file EventStorage (Marten.Events.EventGraph)
   at LamarCompiler.CodeFileExtensions.InitializeSynchronously(ICodeFile file, GenerationRules rules, ICodeFileCollection parent, IServiceProvider services)
   at Marten.Internal.ProviderGraph.StorageFor[T]()
   at Marten.Internal.Sessions.QuerySession.StorageFor[T]()
   at Marten.Internal.Sessions.QuerySession.StorageFinder`1.Find(QuerySession session)
   at Marten.Internal.Sessions.QuerySession.StorageFor(Type documentType)
   at Marten.Linq.Parsing.LinqHandlerBuilder..ctor(MartenLinqQueryProvider provider, IMartenSession session, Expression expression, ResultOperatorBase additionalOperator, Boolean forCompiled)
   at Marten.Linq.MartenLinqQueryProvider.Execute[TResult](Expression expression)
   at Remotion.Linq.QueryableBase`1.GetEnumerator()
   at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
Call to session.Events.QueryAllRawEvents() using IDocumentStore.LightweightSession
...
```

We have tried different attempts to get `EventGraph` generated, but so far in vain (we can see this class in the Marten codebase, but it seems like the code is expecting this to also be pre-built. Is there some way we can force this class to be pre-built also?

I think it will take some time to create a minimal reproducible example, but we are also considering just handrolling our own classes for fetching and persisting events (given that we use so little of the Marten functionality).

Thanks in advance for any help and thanks for a framework that has helped us for many years 👍 

[EventStorage.zip](https://github.com/JasperFx/marten/files/10732154/EventStorage.zip)

**Comment by jaredwri** [comment]
Preface: I've recently had to accomplish the exact same thing for similar reasons (upgrade from `3.14.x` to `5.11.0`).  I'm not an expert, just trying to help.  

I'm using marten in micro-services and have a small runtime memory threshold.  Marten's dynamic compile far exceeds the runtime expectations of my service and does not release the memory as far as I've noticed.  The mechanism for pre-building types does work well for me though and it was pretty simple to get running.

Per my understanding of the documentation you referenced, there were just a few things I needed to do.

1. add a nuget reference to the command line processor `Install-Package Marten.CommandLine -version 5.11.0`.
2. During your host builder processing, add the following.
```csharp
var webHost = Host
      .CreateDefaultBuilder(args)
      // other...
      // Add this
      .ApplyOaktonExtensions()
      .Build();
                
// and this
await webHost.RunOaktonCommands(args);

```
3. In the AddMarten extension

```csharp
services.AddMarten(options =>
{
//...whatever.
opts.AutoCreateSchemaObjects = LamarCodeGeneration.AutoCreate.All;
options.GeneratedCodeMode = LamarCodeGeneration.TypeLoadMode.Auto;

// all pre-built types, queries, and projections need to be registered.  I didn't have to define any events that I used, 
// just the types that represented the document stored in the database. 
})
// I don't know if this needed to be done in your scenario. I'm using telepresence and wanted to make sure it treated this environment as local development.
.OptimizeArtifactWorkflow(TypeLoadMode.Dynamic, "telepresence")
.OptimizeArtifactWorkflow(TypeLoadMode.Dynamic, "development")
.OptimizeArtifactWorkflow(TypeLoadMode.Static)

```


I'm using docker build to create an image that will be deployed to a kubernetes cluster.
1. In my csproj file I added the following item group.
```xml
  <ItemGroup>
    <Folder Include="Internal\Generated\*" />
  </ItemGroup>
```
2. In my docker file I invoked `dotnet run -- codegen delete` and then `dotnet run -- codegen write`.


After these steps I was able to deploy my image which did use the pre-generated files and reduced startup cost and memory usage drastically.
